src/odemis/acq/millmng.py

```diff
@@ -0,0 +1,504 @@
+        else:
+            milling_time = self._milling_time_estimate * remaining_sites
+
+        self._remaining_stage_time -= self.estimate_stage_movement_time(self.sites[site_idx])
+        total_duration = milling_time + self._remaining_stage_time
+
+        return total_duration
+
+    def estimate_stage_movement_time(self, site):
+        """
+        Estimation the time taken by the stage to move from current position to the location of given site
+        :param site: (CryoFeature) consists X,Y,Z coordinates for stage location in m.
+        :return: (float) time to move the stage between two points in s.
+        """
+        # current position from x,y,z of stage position and eliminating rx,ry,rz
+        current_pos = [p[1] for p in self._stage.position.value.items() if len(p[0]) == 1]
+        target_pos = site.pos.value  # list
+        diff = [abs(target - current) for target, current in zip(target_pos, current_pos)]
+        stage_time = math.hypot(diff[0], diff[1], diff[2]) / self._move_speed
+
+        return stage_time
+
+    def estimate_drift_time(self, setting):
+        """
+
+        return (float): estimated time to acquire 1 anchor area
+        """
+        # nb_anchor_scanning = numpy.ceil(
+        #     self._settings[setting_nb]["duration"] / self._settings[setting_nb]["dcPeriod"])  # class , va's
+        drift_acq_time = 0
+        if setting.dcRoi.value != UNDEFINED_ROI:
+            nb_anchor_scanning = numpy.ceil(setting.duration.value / setting.dcPeriod.value)  # repeat?
+            drift_estimation = AnchoredEstimator(self._scanner, self._detector,
+                                                 setting.dcRoi.value, setting.dcDwellTime.value, max_pixels=512 ** 2, follow_drift=False)
+            drift_acq_time = drift_estimation.estimateAcquisitionTime() * nb_anchor_scanning
+
+        return drift_acq_time
+
+    def run(self):
+        """
+        The main function of the task class, which will be called by the future asynchronously
+
+        """
+        if not self._future:
+            return
+        self._future._task_state = RUNNING
+
+        try:
+            actual_time_per_site = None
+            # if no site is provided
+            if len(self.sites) == 0:
+                logging.debug("Site location is not provided by the user, cancelling the milling")
+                self._future._task_state = CANCELLED
+                raise CancelledError()
+            for site_idx, site in enumerate(self.sites):
+            # for/ site_idx in range(len(self.sites)):
+            #     site = self.sites[site_idx]
+                # Update progress on the milling sites left
+                start_time = time.time()
+                remaining_t = self.estimate_milling_time(site_idx, actual_time_per_site)
+                self._future.set_end_time(time.time() + remaining_t)
+
+                # .sites[idx].status already instantiated as FEATURE_ACTIVE
```

Comment:
We don't really know what the .status is, but that doesn't matter. => drop this comment.